docs/tech/rfc/0019-240310-cesium-race-conditions.md

```diff
@@ -7,129 +7,558 @@
+writing to it. This means that no writer can write to a file currently being garbage collected.
+
+In the index, since GC only modifies pointers on the file being collected, it will not
+change any pointers being written to by the writer, as it is necessarily not on the file,
+for reasons mentioned above.
+
+Therefore, these conflicts are taken care of by the file controller and the mutex, and
+there are no additional actions needed.
+
+### 4.1.4 GC-Delete Contention
+
+The deleter does not modify the data in the file system, so we will only discuss the index.
+Unlike the writer, the deleter may perform operations on the pointers currently being
+garbage collected.
+
+The conflict arises when both entities attempt to change the offset of a file: delete
+must change this to account for deletion after the start of a domain to only take in
+a part of the data; GC must change this to store the new position of the data in the
+file system. Thankfully, both changes to the offset are with respect to the existing value:
+both operations change the offset by adding (delete) or subtracting (GC) some bytes
+to the existing offset. Therefore, it does not matter which one happens first. This way,
+GC and delete can work in tandem, so long as the modifications happen one by one and not
+at the same time, which is handled by the mutex on the index.
+
+### 4.1.5 GC-GC Contention
+
+If two garbage collection subprocesses run at the same time, we risk a serious data race:
+the underlying data in the file system may be modified by two entities at the same time,
+and the offset will be in a bad state. Therefore, we must disallow GC-GC Contention.
+
+### 4.1.6 Delete-Delete Contention
+
+There are two cases for delete-delete contention: either the two time ranges contend over
+the same pointer(s), or they are disjoint: the case where they are disjoint is simple: we
+can simply use Mutex Locks to turn the operations into serial operations to the index and
+there would be no conflicts (see the detailed deletion section 4.2.3).
+
+If the two deletions contend over the same pointers, there are two more cases: the two
+time ranges could overlap each other, in which case the controller at the Unary level would
+return an error and disallow the deletion, or they don't overlap each other, in which case
+the operation is turned into a Serial operation of the index (see section 4.2.3).
+
+The following diagram summarizes possible conflicts between these entities:
+
+<div style="text-align: center">
+    <img src="img/0019-240529-cesium-race-conditions/Entity-Interactions.png">
+    <h6>Cesium entities' interaction</h6>
+</div>
+
+## 4.2 Domain level race conditions
+
+We will now inspect, from the bottom up, how the various entities interact with the
+domain data and the index in memory.
+
+### 4.2.1 Writing
+
+A write proceeds in three phases:
+1. Acquire a writer (getting a file handle to an existing file or opening a new one)
```

Comment:
What is the process of acquiring a writer? Aren't there potential race conditions there?

Reply:
I see you added more detail later